2-Draft-Accepted/RFCNNNN-Support-Experimental-Features.md

```diff
@@ -0,0 +1,626 @@
+- [`Compiler.GenerateTypesAndUsings(ScriptBlockAst rootForDefiningTypesAndUsings, List<Expression> exprs)`](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/parser/Compiler.cs#L2199)
+- [`PSModuleInfo.CreateExportedTypeDefinitions(ScriptBlockAst moduleContentScriptBlockAsts)`](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/Modules/PSModuleInfo.cs#L612)
+- [`CompletionCompleters.CompleteType(CompletionContext context, string prefix = "", string suffix = "")`](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs#L5844)
+
+The searching for `TypeDefinitionAst` nodes should be done in one method,
+so that there is no duplication of the filtering logic for the `[Experimental()]` attribute.
+
+As for the type members with the `[Experimental()]` attributes,
+[`TypeDefiner`](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/parser/PSType.cs#L17) needs to be updated to process or skip a type member when processing a `TypeDefinitionAst` node,
+depending on whether the experimental feature is enabled and the `ExperimentAction` value.
+
+#### Exposed Public APIs
+
+An experimental feature may expose public C# APIs.
+Public C# APIs cannot be hidden.
+Even if we alter the `DotNetAdapter` to filter out those public APIs from the experimental feature in PowerShell,
+they will still be visible to applications that host PowerShell.
+Therefore, we will not do anything about the public APIs exposed from experimental features.
+
+### Experimental Feature Discovery
+
+A user should be able to discover the available experimental features in PowerShell,
+including the ones in the PowerShell engine as well as the ones from modules.
+
+#### PowerShell Engine Experimental Feature
+
+For experimental features in PowerShell engine,
+an internal static read-only collection `EngineExperimentalFeatures` will be used to track all available engine experimental features.
+The items in the collection are of the type `ExperimentalFeature`, whose definition looks as follows:
+
+```c#
+public class ExperimentalFeature
+{
+    // Feature name
+    public string Name { get; }
+    // Feature description
+    public string Description { get; }
+    // Feature source (e.g. PSEngine, or module name)
+    public string Source { get; }
+    // Feature is enabled or not
+    public bool IsEnabled { get; }
+}
+
+internal static readonly ReadOnlyCollection<ExperimentalFeature> EngineExperimentalFeatures;
+```
+
+When adding a new experimental feature, an instance representing the feature needs to be added to the collection for it to be discoverable to users.
+The `Source` for engine experimental features should be specified as `PSEngine`.
```

Comment:
Is this always true?  If so then why do we need a settable 'Source' property?

Reply:
Modules can expose experimental features, and for those, the 'Source' would be the modules names. Therefore, the 'Source' property needs to be settable.